src/Data/Text/Utf8/AhoCorasick/Automaton.hs

```diff
@@ -243,13 +239,25 @@ type ValuesMap v = IntMap [v]
 
 -- | Build the trie of the Aho-Corasick state machine for all input needles.
 buildTransitionMap :: forall v. [([CodeUnit], v)] -> (Int, TransitionMap, ValuesMap v)
-buildTransitionMap =
+buildTransitionMap needles = buildTransitionMap' [(decodeUtf8 cus, val) | (cus, val) <- needles]
+
+-- | Decode a list of UTF-8 code units into a list of code points.
+decodeUtf8 :: [CodeUnit] -> [CodePoint]
+decodeUtf8 [] = []
+decodeUtf8 (cu0 : cus) | cu0 < 0xc0 = fromIntegral cu0 : decodeUtf8 cus
+decodeUtf8 (cu0 : cu1 : cus) | cu0 < 0xe0 = Utf8.decode2 cu0 cu1 : decodeUtf8 cus
+decodeUtf8 (cu0 : cu1 : cu2 : cus) | cu0 < 0xf0 = Utf8.decode3 cu0 cu1 cu2 : decodeUtf8 cus
+decodeUtf8 (cu0 : cu1 : cu2 : cu3 : cus) = Utf8.decode4 cu0 cu1 cu2 cu3 : decodeUtf8 cus
+decodeUtf8 cus = error $ "Invalid UTF-8 input sequence at " ++ show (take 4 cus)
```

Comment:
This looks a bit inefficient. We don't need to do anything about this now, but it'd be good to revisit building performance as well, as we ran into edge cases with that before, and this doesn't appear to make it faster.

Reply:
I agree, got any ideas here yet?

Reply:
Nothing concrete, let's leave it like this for now, and come back later for some profiling of this part of the code.